src/shell/output.c

```diff
@@ -586,37 +586,84 @@ static int shell_output_parse_type (struct shell_output *out,
     return 0;
 }
 
-/* handle mustache templates and the similar special cases */
-static char *shell_output_get_path (struct shell_output *out, const char *path)
-{
-    char *rv = NULL;
-    char *ptr;
-
-    /* replace {{id}} with jobid */
-    if ((ptr = strstr (path, "{{id}}"))) {
-        char buf[32];
-        int len, buflen, total_len;
-
-        len = strlen (path);
-        buflen = snprintf (buf, sizeof (buf), "%ju",
-                           (uintmax_t)out->shell->info->jobid);
-        /* -6 for {{id}}, +1 for NUL */
-        total_len = len - 6 + buflen + 1;
-        if (total_len > (PATH_MAX + 1)) {
-            errno = EOVERFLOW;
-            return NULL;
+static int fwrite_shell_info (flux_shell_t *shell, FILE *fp, const char *name)
+{
+    char value[128];
+
+    /*  "jobid" is a synonym for "id" */
+    if (strncmp (name, "jobid", 5) == 0)
+        name += 3;
+    if (strncmp (name, "id", 2) == 0) {
+        const char *type = "f58";
+        if (strlen (name) > 2) {
+            if (name[2] != '.') {
+                shell_log_error ("bad template name %s", name);
+                return -1;
+            }
+            type = name+3;
+        }
+        if (flux_job_id_encode (shell->info->jobid,
+                                type,
+                                value,
+                                sizeof (value)) < 0) {
+            shell_log_errno ("flux_job_id_encode failed for %s", name);
+            return -1;
         }
-        if (!(rv = calloc (1, total_len)))
-            return NULL;
-        memcpy (rv, path, ptr - path);
-        memcpy (rv + (ptr - path), buf, buflen);
-        memcpy (rv + (ptr - path) + buflen, ptr + 6, len - (ptr - path) - 6);
     }
-    else {
-        if (!(rv = strdup (path)))
-            return NULL;
+    return fputs (value, fp);
+}
+
+static char *mustache_render (struct shell_output *out, const char *template)
+{
+    char name [1024];
+    size_t size;
+    char *result = NULL;
+    const char *pos = template;
+    FILE *fp = open_memstream (&result, &size);
+
+    if (fp == NULL) {
+        shell_log_errno ("open_memstream");
+        goto fail;
+    }
+    for (;;) {
+        int len;
+        char *end;
+        char *start = strstr (pos, "{{");
+        if (start == NULL) {
+            if (fputs (pos, fp) < 0) {
+                shell_log_errno ("fputs(%s)", pos);
+                goto fail;
+            }
+            break;
+        }
+        if (fwrite (pos, start - pos, 1, fp) == 0) {
```

Comment:
I saw some `flux-shell: ERROR: output: output.c:640: fwrite: Invalid argument` errors.  I think its b/c if the template begins with `{{`, then `start - pos` == `0`.  So this `fwrite()` call should only be done if `start != pos`.

Reply:
Yeah, hm. I had that in there at first, but I thought I tested that `fwrite` with `size == 0` would just silently write 0 bytes. Guess not. I'll restore the check for `start > pos`. Thanks, good catch!